google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsBatchingDescriptorV2.java

```diff
@@ -0,0 +1,127 @@
+ * limitations under the License.
+ */
+package com.google.cloud.bigtable.data.v2.stub.mutaterows;
+
+import com.google.api.core.InternalApi;
+import com.google.api.core.SettableApiFuture;
+import com.google.api.gax.batching.BatchingDescriptor;
+import com.google.api.gax.batching.BatchingRequestBuilder;
+import com.google.api.gax.rpc.ApiException;
+import com.google.bigtable.v2.MutateRowsRequest;
+import com.google.cloud.bigtable.data.v2.models.BulkMutation;
+import com.google.cloud.bigtable.data.v2.models.MutateRowsException;
+import com.google.cloud.bigtable.data.v2.models.MutateRowsException.FailedMutation;
+import com.google.cloud.bigtable.data.v2.models.Mutation;
+import com.google.cloud.bigtable.data.v2.models.RowMutationEntry;
+import com.google.common.base.Function;
+import com.google.common.collect.Maps;
+import com.google.protobuf.ByteString;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * A custom implementation of a {@link BatchingDescriptor} to split Rpc response into individual
+ * rows response and in a {@link MutateRowsException}.
+ *
+ * <p>This class is considered an internal implementation detail and not meant to be used by
+ * applications directly.
+ */
+// TODO(rahulkql): Consider renaming this class
+@InternalApi("For internal use only")
+public class MutateRowsBatchingDescriptorV2
+    implements BatchingDescriptor<RowMutationEntry, Void, BulkMutation, Void> {
+
+  @Override
+  public BatchingRequestBuilder<RowMutationEntry, BulkMutation> newRequestBuilder(
+      BulkMutation prototype) {
+    return new MyRequestBuilder(prototype);
+  }
+
+  @Override
+  public void splitResponse(Void som, List<SettableApiFuture<Void>> batch) {
+    for (SettableApiFuture<Void> batchResponse : batch) {
+      batchResponse.set(null);
+    }
+  }
+
+  @Override
+  public void splitException(Throwable throwable, List<SettableApiFuture<Void>> batch) {
+    if (!(throwable instanceof MutateRowsException)) {
+      for (SettableApiFuture<Void> future : batch) {
+        future.setException(throwable);
+      }
+      return;
+    }
+
+    List<FailedMutation> failedMutations = ((MutateRowsException) throwable).getFailedMutations();
+
+    Map<Integer, FailedMutation> errorsByIndex =
+        Maps.uniqueIndex(
+            failedMutations,
+            new Function<FailedMutation, Integer>() {
+              @Override
+              public Integer apply(@Nullable FailedMutation input) {
+                return input.getIndex();
+              }
+            });
+
+    int i = 0;
+    for (SettableApiFuture<Void> response : batch) {
+      // NOTE: The gax batching api doesn't allow for a single response to get different
+      // exception for different entries. However this does not affect this client because
+      // SettableApiFuture corresponds to  single element. So just use the last error per entry.
+      ApiException lastError = null;
```

Comment:
I dont understand what you mean here. How can a single entry get multiple errors?

Reply:
I was trying to map FailedMutation to each entry, I had assumed that `MutateRowsException#getFailedMutation()` could return different number of FailedMutation than the number of entries on that batch.

In case the number of entries exceeds the number of FailedMutation, we are setting the last failedMutation to the rest of the entries.

Should we set throwable to each entry (without checking MutateRowsException)? Something like this:
```java
  ...
  @Override
  public void splitException(Throwable throwable, List<SettableApiFuture<Void>> batch) {
    for (SettableApiFuture<Void> future : batch) {
      future.setException(throwable);
    }
  }
  ...
```

Reply:
The first approach is correct (unpacking the entry errors from `MutateRowsException#getFailedMutations`). However the comment is misleading. Each entry will have at most 1 error. In fact if there were duplicate keys then `Maps.uniqueIndex` would throw an IllegalArgumentException. To be honest I think this would be easier to read:

```
Map<Integer, Throwable> entryErrors = new HashMap<>();
for(FailedMutation failure : failedMutations) {
  entryErrors.put(failure.getIndex(), failure.getError();
}

int i = 0;
for (SettableApiFuture<Void> entryResultFuture : batch) {
  Throwable entryError = entryErrors.get(i++);
  if (entryError == null) {
     entryResultFuture.set(null);
  } else {
     entryResultFuture.setException(entryError);
  }
}    
```

Reply:
Thanks a lot for simplifying it. I have updated this method with the same approach.